File: datasets/Part 7 - Natural Language Processing/natural_language_processing.py
# -*- coding: utf-8 -*-
"""
Created on Sat May 21 18:50:24 2022

@author: mudar
"""

# Natural Language Processing

# Importar las librerias
import numpy as np
import matplotlib.pyplot as plt
import pandas as pd

# Importar el dataset
dataset = pd.read_csv('data/Restaurant_Reviews.tsv',
                      delimiter = '\t',
                      quoting = 3) # Ignora las comillas dobles

# Limpieza de texto
import re
import nltk
nltk.download('stopwords') # Eliminar palabras irrelevantes
from nltk.corpus import stopwords
from nltk.stem.porter import PorterStemmer

corpus = []

for i in range(0, 1000):
    # Sustituir por un espacio todos los caracteres excepto
    # con las letras de a la a a la z minusculas y mayusculas
    review = re.sub('[^a-zA-Z]', ' ', dataset['Review'][i])
    review = review.lower() # Pasar a minusculas
    review = review.split() # Pasar a lista la review
    ps = PorterStemmer()
    review = [ps.stem(word) for word in review if not word in set(stopwords.words('english'))]
    review = ' '.join(review)
    corpus.append(review)

# Crear el Bag of Words
from sklearn.feature_extraction.text import CountVectorizer
cv = CountVectorizer(max_features = 1500)
X = cv.fit_transform(corpus).toarray()
y = dataset.iloc[:, 1].values

# Dividir el dataset en conjunto de entrenamiento y conjunto de testing
# 80% conjunto entrenamiento y 20% conjunto de testing
from sklearn.model_selection import train_test_split
X_train, X_test, y_train, y_test = train_test_split(X, y, test_size=0.2, random_state=0)

# Ajustar el clasificador en el Conjunto de Entrenamiento
from sklearn.naive_bayes import GaussianNB
classifier = GaussianNB()
classifier.fit(X_train, y_train)

# Prediccion de los resultados con el Conjunto de Testing
y_pred = classifier.predict(X_test)

# Elaborar una matriz de confusion
from sklearn.metrics import confusion_matrix
cm = confusion_matrix(y_test, y_pred)

print(cm)
(55+91)/200

File: datasets/Part 7 - Natural Language Processing/natural_language_processing.R

# Natural Language Processing

setwd("~/repos/machinelearning-az/datasets/Part 7 - Natural Language Processing")

# Importar el dataset
dataset_original = read.delim('data/Restaurant_Reviews.tsv', quote = '',
                     stringsAsFactors = FALSE)

# Limpieza de textos
library(tm) # Instalar con install.packages('tm')
library(SnowballC) # Instalar con install.packages('SnowballC')
corpus = VCorpus(VectorSource(dataset_original$Review))
corpus = tm_map(corpus, content_transformer(tolower)) # Convertir a minuscula
# Consultar el primer elemento del corpus
#as.character(corpus[[1]])
corpus = tm_map(corpus, removeNumbers) # Eliminar numeros de las reviews del corpus
corpus = tm_map(corpus, removePunctuation) # Eliminar signos de puntuacion
corpus = tm_map(corpus, removeWords, stopwords(kind = 'en')) # Eliminar palabras irrelevantes
corpus = tm_map(corpus, stemDocument) # Eliminar las "variantes" de palabras por su "raiz" ej: loved -> love
corpus = tm_map(corpus, stripWhitespace) # Eliminar espacios en blanco

# Crear el modelo Bag of Words
dtm = DocumentTermMatrix(corpus)
dtm = removeSparseTerms(dtm, 0.999)

dataset = as.data.frame(as.matrix(dtm))
dataset$Liked = dataset_original$Liked

# Codificar la variable de clasificacion como factor
dataset$Liked = factor(dataset$Liked, levels = c(0, 1))

# Dividir los datos en conjunto de entrenamiento y conjunto de test
library(caTools) # Instalar con install.packages('caTools')
set.seed(123) # Para que salgan los mismos resultados que en el curso
split = sample.split(dataset$Liked, SplitRatio = 0.80)
training_set = subset(dataset, split == TRUE)
testing_set = subset(dataset, split == FALSE)

# Ajustar el clasificador de Random Forest con el conjunto de entrenamiento
library(randomForest) # Instalar con install.packages('randomForest')
classifier = randomForest(x = training_set[, -692],
                          y = training_set$Liked,
                          ntree = 10)

# Prediccion de los resultados con el conjunto de Testing
y_pred = predict(classifier, newdata = testing_set[, -692]) # Todo el dataset excluyendo la columna 3

# Crear la matriz de confusion
cm = table(testing_set[, 692], y_pred)
